### DISCUSSION TITLE
Adobe air future

### DISCUSSION THREAD

**Discussion by 123uyt** [airsdk/Adobe-Runtime-Support#1257] [General]
I really want to stay with AIR and try making a professional game in the upcoming months using this, however I have been extremely troubled as I can't find a well put together documentation on air.

Unless the one provided by adobe is the only one out there, which means harman added nothing new. It is practically impossible to see what you guys are doing and what is being improved. Maybe I am wrong, please link me the new documentation. That being said there are a few recommendations I have that would be critical to game development going forward in future. 

1. Add support for websockets
2. Improve how movieclips work, give us more control, this is probably the most important because most complex animations are movieclips, especially if such things require code. When a movieclip plays, each frame is never 'cached'. It literally costs memory to run a movieclip over and over. 
3. I assume 'cache as bitmap' isn't readily available for desktops because the platform wants people to move away from vector, would have been a nice feature but can't complain much on that.

**Comment by Jason0247** [comment]
@ajwfrost 
Indeed, we now lack a complete document to refer to development.

In the aspect of MovieClip, when you play it once and then play it a second time, the speed will be very fast. I don't know if there is any way to cache in advance, which is very important.

In fact, I think there is a way to reduce the workload of Harman team and effectively solve the problems of developers.

It is to establish a problem community, classify some common, practical and widely used problems, and then the Harman team will provide solutions, preferably examples. Then make such a section, and you don't have to do so complete official documents as Adobe did before. In fact, for air developers, everyone is distressed by these problems, including the current discussion area, which are also some common difficulties and complications. It seems that you don't need complete documents.

**Comment by Jason0247** [comment]
In addition, I don't think some low-level problems should still be solved by the Harman team after they appear. Our developers should also help each other solve them and take the initiative to give answers, so as not to let these useless problems affect the whole ecology. For example, how to control the pause and playback of MovieClip

In fact, what I saw was that Andrew frost answered almost all the questions and answered them carefully. Thank you for that.

**Comment by ajwfrost** [comment]
@123uyt documentation is probably one of our weak points .. the changes we've been making are all listed in the release notes (either from the [website](https://airsdk.harman.com/release_notes) or as a long list within the [PDF](https://airsdk.harman.com/api/versions/33.1.1.633/release-notes/Release_Notes_AIR_SDK_33.1.1.633.pdf).

But the goal is to increase the documentation and have user-submitted content as well, via the [new AIR SDK developer portal](https://airsdk.dev/).

In terms of your feature requests:

1) WebSocket support as a server? or as a client? (or both?!) - this probably wouldn't be a high priority since this capability could be added purely in ActionScript e.g. see https://github.com/theturtle32/AS3WebSocket

2) I think this could be quite complex if anyone had nested movieclips or any dynamic content within a movieclip. I believe that all the layouts and information regarding the movieclip contents would be cached as long as nothing is done that would make this flagged as dirty, but it's an area where we'd risk some regression. Plus of course, cache as bitmap has some memory constraints! You can potentially do this directly by drawing the movieclip into a BitmapData object?

3) This may be more due to the rendering modes, but I wasn't aware that cache as bitmap didn't work on desktops(?!) - there are some interesting points about this in a [StackOverflow article](https://stackoverflow.com/questions/38340965/as3-cacheasbitmap-confusion). Vector drawing is still a good plan for content that needs to be scaled and the software rendering engine within AIR (cpu/direct modes) is very highly optimised for the SWF shape formats. But it didn't really work very well when Adobe tried to do the rendering via gpu (which are more intended for 3D graphics; OpenVG didn't really go anywhere sadly...)


thanks

**Reply by 123uyt** [reply]
Thanks for comment, I think websocket support would be the game changer for air. The link you provided will not work with adobe animate IDE as that isn't flex. Also note worthy that was made 8 years ago.

**Reply by ajwfrost** [reply]
The library there was written in pure AS3, although compiled with Flash Builder and the test app is written using Flex. But it should be possible to link the .swc within Animate. Regarding the age of the library, I'm not sure whether the WebSockets standard will have changed that much in 8 years but yes there may be some newer capabilities that need to be added to it..
thanks

**Reply by 123uyt** [reply]
I believe adobe air secure socket (tls) could work out for me. A few problems I ran into recently is, I can't connect to any websockets, either local, or simulated sites that allow you to connect to theirs for testing purposes. The only websocket so far, that I was able to connect to was help.adobe.com. Maybe outdated?

**Reply by RayGyoe** [reply]
@ajwfrost  On the third point, whether to further improve the rendering performance of bitmap on the basis of classical rendering?

Our video calling software uses ane to return bytearray and draw it to bitmapData. With more than 20 callers and 24fps, the performance drops sharply

**Reply by ajwfrost** [reply]
@RayGyoe  there are some specific workflows that probably can be optimized a lot ... what's in your byte array, is it just pixel data i.e. do you mean you call `BitmapData.setPixels` with it? Is there any format conversion already going on that we could eliminate e.g. do you get the initial images/video frames like that, or is it a YUV format or similar? as there are already a lot of optimized rendering routines inside the runtime that we could use for this.

AIR has to make assumptions about what you will do with the data so will be copying everything, potentially with the ability to do scaling/smoothing/etc, but it would possibly work better if we could have a function where the BitmapData takes ownership of the byte array and doesn't have to copy it. If you can clarify the workflow here, we can look to see what's possible.

thanks

**Reply by RayGyoe** [reply]
@ajwfrost yes, get RGBA bytearray call BitmapData.setPixels updates the image.
the performance loss is not at the C++ level. We also tried to use stage3d, and tested that there was no faster or lower CPU consumption than classical rendering on some mobile notebooks (especially the old amd graphics card stage3d achieved worse performance)

To this end, we tried QT / WPF / AIR / haxe demo
Test benchmark: 50 images refresh RGBA byte at 24fps
Prototype: CPU i9 / Intel UHD + n T1000 / 4K display
Performance: 
QT/WPF 60fps (cpu ≈15%)
AIR 24fps (cpu ≈25%)
Haxe 10fps (cpu ≈40%)

**Reply by ajwfrost** [reply]
@RayGyoe if you have a test/benchmark app, would it be possible for us to get hold of this and try it out? Would love to see if we can optimize this..

**Reply by RayGyoe** [reply]
@ajwfrost  thank you for your attention, I will send it after finishing

**Comment by joebillman** [comment]
I have had this same question. As someone that initially started his programming career with ActionScript 1.0 I absolutely love ActionScript and AIR. That said, it is really sad to see where things have ended up. I would love to see AIR succeed, but it feels like Harman doesn't plan to keep it going at the level it deserves. Back in the day I was part of a team that created a large scale SAAS product built on Flex. As part of that we created our own language reference that was similar to the AS3 reference and used the tool provided by the framework for creating language references. I've always loved the AS3 language reference (documentation) and using that tool made it pretty simple to create. It seems to me like the team at Harman could totally host an updated version of the API. The fact that they aren't makes me think that they aren't really committed to improving AIR/ActionScript and moving it forward. I wish they were :(

**Reply by MalacTheLittle** [reply]
If you would still be in love with AIR, you would notice that after Harman took it over, AIR has gotten wings! Obviously you're not, since you don't know about all the struggle they are making to keep up with Stores and OS's imposed restrictions. And in all that struggle they find place to improve some long forgotten gems.
And while theres Harman with enthusiasts like ajwfrost and his team, I am sure AIR comunity will be getting bigger!

**Reply by joebillman** [reply]
Fair point, I didn't mean any disrespect, just opinions/observations. After posting today, I took some time to look over the latest release notes and noticed that they are regularly updating the sdk which is promising. Maybe I spoke out of line in regards to Harman. I hope you're right, that it will keep getting bigger. It's just hard to see where it is now after having been around during the glory days of Flash, Flex, and ActionScript. I mostly develop in JavaScript and TypeScript now and miss ActionScript a lot. I must say that developing games for iOS using Starling was one of the funniest things I've done during my 16+ years as a programmer. Here's to AIR's comeback!

**Reply by marchbold** [reply]
Regarding the reference docs, I know they are working on that. I'm not exactly sure of the details but from what I've been told it appears to be Adobe holding it up though as the docs are generated from a different process than the runtime Harman have licensed. so they are working out how best to generate them. 

Once they have I believe they have plans to host them here: https://airsdk.dev/reference

**Reply by MalacTheLittle** [reply]
@joebillman No hard feelings :)
I just wanted to point out that all these neglecting years of AIR and Flash has taken a tribute, so it needs some time to rise from the ash. And since it's not just Adobe's fault but also up to all the glorification of html and js and forcement of native languages, it needs **SOME** time to come back.
And of course, you're always welcome back :) I'm sure you would enjoy it :)

**Reply by joebillman** [reply]
I appreciate the info from both of you. I really do hope that there is a great comeback of AIR! The company I work for was forced to move all of our stuff over to TypeScript and HTML5 years ago from ActionScript and Flex because we need it to run natively in the browser, but I would love to do some freelance stuff in AS3 and AIR again.

**Reply by joebillman** [reply]
Maybe there's still hope for AS4, that would be awesome! I saw some of the initial concepts of that back in the day from some of the Flash Player devs and it looked pretty cool. Looked similar to how Swift looks now, but it was ActionScript, so it was better :)

**Reply by ajwfrost** [reply]
Thanks for the input - yes we've been planning to take over the documentation website so that we can tweak it for the API changes that we've made. Various different issues holding it up...

It's always useful for us to get feedback, and interesting that the documentation is seen as such a key part. I know @marchbold has the same feeling (as he's told us so, several times!) and with his help on the https://airsdk.dev website, we should have a place where we can start to build it all up. Community contributions are very welcome!

AS4 would be a massive step change: this would have been a completely new virtual machine (which would then have had to been integrated into AIR parallel to avmplus) so although we had some visibility of this when it was first touted, we are not planning to do something quite so radical. There are some gradual AS3 improvements we're hoping to make though... and of course, the avmplus itself and the ABC format are openly available in case anyone feels like creating a new ActionScript version with new features that compiles back down to ABC to run in the existing avmplus engine... :-)

Going back to the original point ... it may take a bit of time but hopefully google searches will start to find content on the new AIR developer portal. And we have an open action with our Harman-based website to make it point that way a little clearer (our own website is constrained by our marketing team and branding guidelines, plus I think it's a lot better to have a generic, open-source/CC-licensed github location for all of this documentation so that no one company has the ability to suddenly take it away!)

thanks

**Reply by joebillman** [reply]
@ajwfrost, thank you so much for your comments. All of the interaction with everyone on here has caused me to do a complete 180 from where my feelings were when I posted my original comment. I'm excited again for AIR's future! The quick responses combined with the release history that I reviewed yesterday do remind me of the glory days. Keep up the great work! What was said about AS4 makes total sense. If AS3 continues to get continual improvements, that would be amazing. AS3 is still my favorite programming language BTW. In regards to the documentation, I do think that would be huge if you can host it so that you can update it with the changes that have been made to the framework. If I can help in some way, I'm totally willing to. I have one more question: Does anyone know what the support looks like from the developers of the IDEs? I've used most of them and still prefer Flash Builder, IntelliJ and Visual Studio Code in that order. The work that Josh Tynjala has done on the plugin for VSCode is awesome, but VSCode is still missing some of the power that the other IDEs have. I know that Jetbrains continues to include their Flash/Flex updates in IntelliJ, but I'm curious as to how much effort they can afford to keep putting in. My main question, is regarding Flash Builder. Has there been any discussion about Adobe turning over Flash Builder to Harman? In my opinion Flash Builder is still the best editor for one reason, the profiler. Using that with Adobe Scout was very powerful. That would be cool if they turned over Scout to Harman too. Flash Builder would just need some small improvements, the biggest would be to bring it up-to-date. I know that FB was built on an Eclipse base. I know that it's easier said than done, but even updating that would be huge. Another cool thing would be to update the emulators to some of the more popular devices of today. The emulators in Android Studio and Xcode have come a long way and it would be cool to have something similar. Anyway I'll stop with my Christmas wish list. Thanks again for the comments!

**Reply by ajwfrost** [reply]
@joebillman thanks - yes it's encouraging us too :-) and it would be even better if we can get new folk using it. But as @MalacTheLittle says, we probably have a bit further to go before we can hope for too much..

Regarding Flash Builder ... Adobe stopped supporting it quite a while ago, but it was part of the Creative Cloud offering. So trying to take that over was going to be a little more complex commercially. It's actually a bit of a surprise to us how many people are still using it! So, we don't have any plans to try to take this over I'm afraid. I can check with them about the profiler though. Scout is something we are planning to maintain, there are just some delays in getting it all sorted out..

I hadn't thought about the emulators .. do you mean, the list of options in ADL:
```
PREDEFINED_SCREEN_SIZE : iPhone|iPhoneRetina|iPhone5Retina|iPhone6|iPhone6Plus|iPod|iPodRetina|iPod5Retina|iPad|iPadRetina|Droid|NexusOne|SamsungGalaxyS|SamsungGalaxyTab|QVGA|WQVGA|FWQVGA|HVGA|WVGA|FWVGA|1080|720|480
```

thanks

**Reply by joebillman** [reply]
Yes, those are the emulators I was referring to. Sorry, one other question came to mind. Back in the day there was some discussion around AIR on wearables like the Apple Watch. What are your thoughts on that?

**Reply by ajwfrost** [reply]
I hadn't really paid attention to those emulators, I believe they're just a shortcut for defining the UI sizes so it should be trivial to change those. And perhaps more future-proof to have it read the aliases from a config file somewhere... I've just added a task for this into our backlog, I can't say it would be a massively high priority but equally it's a nice simple task for someone to pick up when they have a bit of time..

For Apple Watch .. we haven't looked into it, afaik you're the first person to mention it! We'd look at new platforms based on the demand and the ease of it (hence bringing back Linux, even though it's a relatively small number of people who wanted it - but it's also a good stepping stone to some other platforms..). For Apple Watch I don't know how many people would be writing apps that would really benefit from this, but I can get one of the team to at least do a quick check as to what might be involved..

thanks

   Andrew

**Reply by joebillman** [reply]
Thanks again Andrew for the quick responses. It is really awesome to see AIR moving forward. Kudos to you and your team(s)! I will be paying more attention and I really am willing to help in the future if I can.

**Reply by ajwfrost** [reply]
I enjoy it too which is always a bonus :-)   I've worked on our Adobe partnership since 2005 putting Flash and Flash Lite into Nokia smartphones, it has to be said that AIR SDK is a very different side to it! Documentation/tutorial contributions - and additions to the new showcase area on the airsdk.dev website - are always welcome!

cheers

**Reply by joebillman** [reply]
I developed some games a while back with AIR and Starling. They are the rights of my employer, but I will see if they are cool with me putting some of that work in the showcase and I'd be happy to put new stuff there after it's developed. I would also be willing to look into tutorials. I've had that thought multiple times over the years. I thought of one more question: Is there much going on with ANEs? I've noticed that most of the people that were building them and offering them have stopped. I was looking into some promising things with creating native extensions using Swift instead of Objective C and that seemed really cool since Swift is much nicer to code. Curious to hear from you what the future thoughts are for ANEs.

**Reply by ajwfrost** [reply]
Sounds good - tutorials or code examples are always very useful!

There are still some folk offering ANEs, the most prevalent being Distriqt but I know others who have some Swift-related ANEs. We've been discussing with Distriqt on how to optimize ANE implementation/integration on Android in particular, as part of our updates to start using Gradle and standard Android toolchains to do the compilation/packaging for that.. for iOS and for desktop platforms there may be improvements we can make but the main focus of effort at the moment is the "AIR Package Manager" that Distriqt are driving, which aims to sort out and simplify a lot of the ANE (and other library) dependency management tasks. See https://github.com/airsdk/apm/wiki

**Comment by al-sabr** [comment]
I would not recommend to use https://github.com/theturtle32/AS3WebSocket because of the encryption library dependency https://github.com/51systems/as3crypto which is not up to date and only support TLS 1.0.

This might be also connected to this thread if the API provides a way to hook OpenSSL encryption containers in AS3 then the problem will be resolved for SSL WebSockets.

https://github.com/airsdk/Adobe-Runtime-Support/issues/1439

**Comment by KupoGames** [comment]
Personally, I've had a much better experience with AIR/Flash since HARMAN took over. 
Issues that Adobe ignored for many years are finally being addressed, and while there's still some bugs, it feels like the AIR devs are actually working on them and listening to my needs.

I've got a few successful Steam games that I'm porting to mobile, and AIR makes it really easy to do so.
For example, I have a shoot-em-up made in Starling, that works almost perfectly on my Galaxy Note8 phone with no changes - the PC version of the game runs smoothly on mobile! (I will need to make some minor changes to improve the UI though)

Even games made with complex vector art run reasonably well on higher-end mobile devices, you can try my RPG for free here: https://play.google.com/store/apps/details?id=air.EpicBattleFantasy5&hl=en&gl=US
To port it to mobile, I basically just had to remove filters, use cacheAsBitmap more, and make the UI bigger - that's all!

Obviously it would be nice if AIR worked on consoles etc, but I think for PC and mobile cross-platform development, it's already great, despite a few limitations. And Distriqt's ANEs provide lots of essential native features too.

So while AIR has not reached it's potential, I still think it's viable game engine, especially if you have an existing portfolio of Flash games.
